Build failure: php-simplexml-8.1.2

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *php-simplexml-8.1.2*

### Build log

```
> PASS SimpleXML: invoking key() after the iterator has already been consumed [tests/key_error.phpt]
┃        > SKIP GHSA-3qrf-m4j2-pcrr (libxml global state entity loader bypass) [tests/libxml_global_state_entity_loader_bypass.phpt] reason: zend-test extension not available
┃        > PASS SimpleXML [profile]: Accessing a simple node [tests/profile01.phpt]
┃        > PASS SimpleXML [profile]: Accessing an array of subnodes [tests/profile02.phpt]
┃        > PASS SimpleXML [profile]: Accessing an attribute [tests/profile03.phpt]
┃        > PASS SimpleXML [profile]: Accessing a namespaced element [tests/profile04.phpt]
┃        > PASS SimpleXML [profile]: Accessing an aliased namespaced element [tests/profile05.phpt]
┃        > PASS SimpleXML [profile]: Accessing a namespaced attribute [tests/profile06.phpt]
┃        > PASS SimpleXML [profile]: Accessing an aliased namespaced attribute [tests/profile07.phpt]
┃        > PASS SimpleXML [profile]: Accessing a namespaced attribute without a namespace [tests/profile08.phpt]
┃        > PASS SimpleXML [profile]: Accessing a namespaced element without a namespace [tests/profile09.phpt]
┃        > PASS SimpleXML [profile]: Accessing two attributes with the same name, but different namespaces [tests/profile10.phpt]
┃        > PASS SimpleXML [profile]: Accessing two elements with the same name, but different namespaces [tests/profile11.phpt]
┃        > PASS SimpleXML [profile]: Accessing namespaced root and non namespaced children [tests/profile12.phpt]
┃        > PASS SimpleXML [profile]: Accessing by namespace prefix [tests/profile13.phpt]
┃        > SKIP SimpleXML [interop]: simplexml_import_dom [tests/simplexml_import_dom.phpt] reason: Required extension missing: dom
┃        > PASS simplexml_load_file() [tests/simplexml_load_file.phpt]
┃        > PASS Incorrectly initialized SimpleXmlElement [tests/simplexml_uninitialized.phpt]
┃        > PASS SPL: SimpleXMLIterator [tests/sxe_001.phpt]
┃        > PASS SPL: SimpleXMLIterator and recursion [tests/sxe_002.phpt]
┃        > PASS SPL: SimpleXMLIterator and getChildren() [tests/sxe_003.phpt]
┃        > PASS SPL: SimpleXMLIterator and overridden iterator methods() [tests/sxe_004.phpt]
┃        > PASS SPL: SimpleXMLIterator and overridden count() [tests/sxe_005.phpt]
┃        > =====================================================================
┃        > TIME END 2024-07-30 16:46:43
┃        >
┃        > =====================================================================
┃        > TEST RESULT SUMMARY
┃        > ---------------------------------------------------------------------
┃        > Exts skipped    :    0
┃        > Exts tested     :   12
┃        > ---------------------------------------------------------------------
┃        >
┃        > Number of tests :  131               127
┃        > Tests skipped   :    4 (  3.1%) --------
┃        > Tests warned    :    0 (  0.0%) (  0.0%)
┃        > Tests failed    :    1 (  0.8%) (  0.8%)
┃        > Tests passed    :  126 ( 96.2%) ( 99.2%)
┃        > ---------------------------------------------------------------------
┃        > Time taken      :    1 seconds
┃        > =====================================================================
┃        >
┃        > =====================================================================
┃        > FAILED TEST SUMMARY
┃        > ---------------------------------------------------------------------
┃        > Bug #79971 (special character is breaking the path in xml function) [tests/bug79971_1.phpt]
┃        > =====================================================================
┃        > make: *** [Makefile:136: test] Error 1
┃        For full logs, run 'nix log /nix/store/qhz26ig0n0d43xapgjwp2rj1nllfi133-php-simplexml-8.1.29.drv'.
```

### Additional context

This happen after latest flake update

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.10.1, NixOS, 24.11 (Vicuna), 24.11.20240727.b73c222`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.18.5`
 - channels(root): `"home-manager-23.11.tar.gz, nixos-23.11"`
 - nixpkgs: `/nix/store/qxf6anli54ij0q1sdlnlgx9hyl658a4v-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/interpreters/php/8.1.nix
+++ b/pkgs/development/interpreters/php/8.1.nix
@@ -5,11 +5,27 @@ let
     version = "8.1.29";
     hash = "sha256-h6YDEyY/L1M/GA5xknLKXkfNmITU7DyTcgGY6v+uCCc=";
     extraPatches = [
-      # Fix build with libxml 2.12+.
+      # Fix build with libxml2 2.12+.
       # Patch from https://github.com/php/php-src/commit/0a39890c967aa57225bb6bdf4821aff7a3a3c082
       (fetchpatch {
         url = "https://github.com/php/php-src/commit/0a39890c967aa57225bb6bdf4821aff7a3a3c082.patch";
         hash = "sha256-HvpTL7aXO9gr4glFdhqUWQPrG8TYTlvbNINq33M3zS0=";
+      })
+      # Fix tests with libxml2 2.12
+      (fetchpatch {
+        url = "https://github.com/php/php-src/commit/061058a9b1bbd90d27d97d79aebcf2b5029767b0.patch";
+        hash = "sha256-0hOlAG+pOYp/gUU0MUMZvzWpgr0ncJi5GB8IeNxxyEU=";
+        excludes = [
+          "NEWS"
+        ];
+      })
+      # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+      # Required for libxml2 2.13 compatibility patch.
+      ./php81-fix-libxml2-2.13-compatibility.patch
+      # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+      (fetchpatch {
+        url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+        hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
       })
     ];
   });

--- a/pkgs/development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+++ b/pkgs/development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
@@ -1,0 +1,28 @@
+diff --git a/ext/libxml/php_libxml.h b/ext/libxml/php_libxml.h
+index a1011f0b17..7a7622c482 100644
+--- a/ext/libxml/php_libxml.h
++++ b/ext/libxml/php_libxml.h
+@@ -119,6 +119,23 @@ PHP_LIBXML_API void php_libxml_shutdown(void);
+ ZEND_TSRMLS_CACHE_EXTERN()
+ #endif
+ 
++#if defined(__clang__)
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_START \
++	_Pragma("clang diagnostic push") \
++	_Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_END \
++	_Pragma("clang diagnostic pop")
++#elif defined(__GNUC__)
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_START \
++	_Pragma("GCC diagnostic push") \
++	_Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_END \
++	_Pragma("GCC diagnostic pop")
++#else
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_START
++# define PHP_LIBXML_IGNORE_DEPRECATIONS_END
++#endif
++
+ /* Other extension may override the global state options, these global options
+  * are copied initially to ctxt->options. Set the options to a known good value.
+  * See libxml2 globals.c and parserInternals.c.

--- a/pkgs/top-level/php-packages.nix
+++ b/pkgs/top-level/php-packages.nix
@@ -383,7 +383,7 @@ in {
             "--enable-dom"
           ];
           # Add a PHP lower version bound constraint to avoid applying the patch on older PHP versions.
-          patches = lib.optionals ((lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.14" && lib.versionAtLeast php.version "8.2.7") || (lib.versions.majorMinor php.version == "8.1" && lib.versionAtLeast php.version "8.1.27")) [
+          patches = lib.optionals ((lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.14" && lib.versionAtLeast php.version "8.2.7") || (lib.versions.majorMinor php.version == "8.1")) [
             # Fix tests with libxml 2.12
             # Part of 8.3.1RC1+, 8.2.14RC1+
             (fetchpatch {
@@ -392,6 +392,14 @@ in {
               excludes = [
                 "NEWS"
               ];
+            })
+          ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.1") [
+            # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+            ../development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+            # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+              hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
             })
           ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
             # Fixes compatibility with libxml2 2.13. Part of 8.3.10RC1+, 8.2.22RC1+
@@ -617,7 +625,24 @@ in {
           configureFlags = [
             "--enable-simplexml"
           ];
-          patches = lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
+          patches = lib.optionals (lib.versions.majorMinor php.version == "8.1") [
+            # Fix tests with libxml2 2.12
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/061058a9b1bbd90d27d97d79aebcf2b5029767b0.patch";
+              hash = "sha256-0hOlAG+pOYp/gUU0MUMZvzWpgr0ncJi5GB8IeNxxyEU=";
+              excludes = [
+                "NEWS"
+              ];
+            })
+            # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+            # Required for libxml2 2.13 compatibility patch.
+            ../development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+            # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+              hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
+            })
+          ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
             # Fixes compatibility with libxml2 2.13. Part of 8.3.10RC1+, 8.2.22RC1+
             (fetchpatch {
               url = "https://github.com/php/php-src/commit/4fe821311cafb18ca8bdf20b9d796c48a13ba552.diff?full_index=1";
@@ -647,7 +672,24 @@ in {
           ];
           doCheck = stdenv.isDarwin;  # TODO: a couple tests still fail on *-linux
           internalDeps = [ php.extensions.session ];
-          patches = lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
+          patches = lib.optionals (lib.versions.majorMinor php.version == "8.1") [
+            # Fix tests with libxml2 2.12
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/061058a9b1bbd90d27d97d79aebcf2b5029767b0.patch";
+              hash = "sha256-0hOlAG+pOYp/gUU0MUMZvzWpgr0ncJi5GB8IeNxxyEU=";
+              excludes = [
+                "NEWS"
+              ];
+            })
+            # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+            # Required for libxml2 2.13 compatibility patch.
+            ../development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+            # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+              hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
+            })
+          ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
             # Fixes compatibility with libxml2 2.13. Part of 8.3.10RC1+, 8.2.22RC1+
             (fetchpatch {
               url = "https://github.com/php/php-src/commit/4fe821311cafb18ca8bdf20b9d796c48a13ba552.diff?full_index=1";
@@ -692,7 +734,24 @@ in {
             "--enable-xml"
           ];
           doCheck = false;
-          patches = lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
+          patches = lib.optionals (lib.versions.majorMinor php.version == "8.1") [
+            # Fix tests with libxml2 2.12
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/061058a9b1bbd90d27d97d79aebcf2b5029767b0.patch";
+              hash = "sha256-0hOlAG+pOYp/gUU0MUMZvzWpgr0ncJi5GB8IeNxxyEU=";
+              excludes = [
+                "NEWS"
+              ];
+            })
+            # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+            # Required for libxml2 2.13 compatibility patch.
+            ../development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+            # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+              hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
+            })
+          ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
             # Fixes compatibility with libxml2 2.13. Part of 8.3.10RC1+, 8.2.22RC1+
             (fetchpatch {
               url = "https://github.com/php/php-src/commit/4fe821311cafb18ca8bdf20b9d796c48a13ba552.diff?full_index=1";
@@ -722,7 +781,24 @@ in {
           configureFlags = [
             "--enable-xmlwriter"
           ];
-          patches = lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
+          patches = lib.optionals (lib.versions.majorMinor php.version == "8.1") [
+            # Fix tests with libxml2 2.12
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/061058a9b1bbd90d27d97d79aebcf2b5029767b0.patch";
+              hash = "sha256-0hOlAG+pOYp/gUU0MUMZvzWpgr0ncJi5GB8IeNxxyEU=";
+              excludes = [
+                "NEWS"
+              ];
+            })
+            # Backport of PHP_LIBXML_IGNORE_DEPRECATIONS_START and PHP_LIBXML_IGNORE_DEPRECATIONS_END
+            # Required for libxml2 2.13 compatibility patch.
+            ../development/interpreters/php/php81-fix-libxml2-2.13-compatibility.patch
+            # Fix build with libxml2 2.13+. Has to be applied after libxml2 2.12 patch.
+            (fetchpatch {
+              url = "https://github.com/php/php-src/commit/9b4f6b09d58a4e54ee60443bf9a8b166852c03e0.patch";
+              hash = "sha256-YC3I0BQi3o3+VmRu/UqpqPpaSC+ekPqzbORTHftbPvY=";
+            })
+          ] ++ lib.optionals (lib.versions.majorMinor php.version == "8.2" && lib.versionOlder php.version "8.2.22") [
             # Fixes compatibility with libxml2 2.13. Part of 8.3.10RC1+, 8.2.22RC1+
             (fetchpatch {
               url = "https://github.com/php/php-src/commit/4fe821311cafb18ca8bdf20b9d796c48a13ba552.diff?full_index=1";

